src/v/cluster/archival/archival_policy.cc

```diff
@@ -358,10 +358,14 @@ ss::future<std::optional<std::error_code>> get_file_range(
         upl->max_timestamp = seek.ts;
     }
     // Recompute content_length based on file offsets
-    vassert(
-      upl->file_offset <= upl->final_file_offset,
-      "Invalid upload candidate {}",
-      upl);
+    if (upl->file_offset > upl->final_file_offset) {
+        // This could potentially happen if the log was truncated after
+        // file_offset is set. In this case the index could become empty which
+        // will trigger the condition above. The operation could be retried
+        // later so throwing makes more sense then the assertion.
+        throw std::runtime_error(
+          fmt_with_ctx(fmt::format, "Invalid upload candidate {}", upl));
+    }
     upl->content_length = upl->final_file_offset - upl->file_offset;
```

Comment:
Q1: (this comment is based on the code comment alone) Is it possible for the log to get truncated and have new batches appended where the above condition still passes but the actual content?

Q2: We should be uploading only below LSO, how we end up trying to create upload candidate with an offset which gets truncated? What type of truncation we are talking about?